corva/models/base.py

```diff
@@ -1,47 +1,106 @@
 from __future__ import annotations
 
 from abc import ABC, abstractmethod
-from typing import Any, List, Optional, TypeVar
+from functools import cached_property
+from typing import Any, Generic, List, Optional, Type, TypeVar, Union
 
 from pydantic import BaseModel, Extra
+from pydantic.generics import GenericModel
 
 from corva.network.api import Api
+from corva.settings import Settings
+from corva.state.redis_adapter import RedisAdapter
 from corva.state.redis_state import RedisState
 
 
+class BaseConfig:
+    allow_population_by_field_name = True
+    arbitrary_types_allowed = True
+    extra = Extra.allow
+    validate_assignment = True
+    keep_untouched = (cached_property,)
+
+
 class BaseEvent(ABC):
     @staticmethod
     @abstractmethod
-    def from_raw_event(event: str, **kwargs) -> BaseEvent:
+    def from_raw_event(event: str) -> Union[List[BaseEvent], BaseEvent]:
         pass
 
 
-class BaseContext(BaseModel):
-    """Stores common data for running a Corva app."""
-
-    class Config:
-        arbitrary_types_allowed = True
-        extra = Extra.allow
-
-    raw_event: str
-    app_key: str
+class BaseData(BaseModel):
+    class Config(BaseConfig):
+        pass
 
-    event: Optional[BaseEvent] = None
-    api: Optional[Api] = None
-    state: Optional[RedisState] = None
-    user_result: Any = None
 
+BaseEventTV = TypeVar('BaseEventTV', bound=BaseEvent)
+BaseDataTV = TypeVar('BaseDataTV', bound=BaseData)
 
-class BaseEventData(BaseModel):
-    class Config:
-        extra = Extra.allow
-        allow_population_by_field_name = True
 
+class BaseContext(GenericModel, Generic[BaseEventTV, BaseDataTV]):
+    """Stores common data for running a Corva app."""
 
-BaseEventDataTV = TypeVar('BaseEventDataTV', bound=BaseEventData)
+    class Config(BaseConfig):
+        pass
 
+    _event: BaseEventTV
+    settings: Settings
 
-class ListEvent(BaseEvent, List[BaseEventDataTV]):
-    """Base class for list events (events that consist of more than one event data)."""
+    user_result: Any = None
 
-    pass
+    # api params
+    api_timeout: Optional[int] = None
+    api_max_retries: Optional[int] = None
+
+    # cache params
+    cache_kwargs: Optional[dict] = None
+    cache_data_cls: Optional[Type[BaseDataTV]] = None
+
+    @property
+    def cache_key(self) -> str:
+        return (
+            f'{self.settings.PROVIDER}/well/{self.event.asset_id}/stream/{self.event.app_stream_id}/'
+            f'{self.settings.APP_KEY}/{self.event.app_connection_id}'
+        )
+
+    @cached_property
+    def event(self) -> BaseEventTV:
+        return self._event
+
+    @cached_property
```

Comment:
it is better to use one Api instance as it uses requests.Session which uses HTTP persistent connection [here](https://requests.readthedocs.io/en/master/user/advanced/)

Reply:
This method does not ensure that one Api instance is used. You need to implement singleton on Api to achieve that. Not cache the result of the method. This is a wrong tool for singleton

Reply:
`Api` class is not a singleton by itself. Really `@cached_property` servers here as a method of writing next code:

```python3
from pydantic import BaseModel, validator
from functools import cached_property
from typing import Optional


class Api:
    pass


# method 1
class Data(BaseModel):
    class Config:
        keep_untouched = (cached_property,)

    @cached_property
    def api(self) -> Api:
        return Api()


# method 2
class Data(BaseModel):
    class Config:
        arbitrary_types_allowed = True

    api_: Optional[Api] = None

    @property
    def api(self) -> Api:
        if self.api_ is not None:
            return self.api_

        self.api_ = Api()
        return self.api_


# method 3
class Data(BaseModel):
    class Config:
        arbitrary_types_allowed = True

    api: Api = None

    @validator('api', pre=True, always=True)
    def set_api(cls, v, values) -> Api:
        # extract Api kwargs from values here and pass to constructor

        return Api()
```

Reply:
I don't understand what are you trying to say? Are these methods - alternative solutions?

Reply:
why not make Api class a singleton?

Reply:
Yes, these methods are some alternatives.

I was trying to say, that using [functools.cached_property](https://docs.python.org/3/library/functools.html#functools.cached_property) saves you from burden of writing complex code with hidden private attributes, that support `@properties`. So, it simplifies code of our BaseContext class.

Simpler example would be:

```python3
from functools import cached_property


class Foo1:
    """cached_property doesnt require you to define and manage private attributes"""

    @cached_property
    def baz(self):
        return []


class Foo2:
    """does the same thing as Foo1"""

    def __init__(self):
        self._baz = []

    @property
    def baz(self):
        return self._baz

    @baz.setter
    def baz(self, value):
        self._baz = value


class Foo3:
    """Fails to achieve the same thing as Foo1"""

    @property  # cant do the same code as in Foo1
    def baz(self):
        return []


foo1 = Foo1()
foo2 = Foo2()
foo3 = Foo3()

foo1.baz.append(1)
foo2.baz.append(1)
foo3.baz.append(1)

assert foo1.baz == [1]
assert foo2.baz == [1]
assert foo3.baz == []


foo1.baz = []
foo2.baz = []

assert foo1.baz == []
assert foo2.baz == []
```

Reply:
> why not make Api class a singleton?

Api class instances can be parametrized by passing different parameter sets to constructor. This parametrization requires different instances to be returned. Singleton would always return only one instance, which was the very first one created.

If we needed to optimize request.Session usage inside Api class, it would be more appropriate to use Flyweight pattern, which would return same instances for same sets of parameters.

These optimizations are not needed at this point in my opinion.